### ISSUE TITLE
Blur effect shader sometimes fails to compile

### ISSUE THREAD

**Issue by gamax92** [gamax92/wine-overwatch#20]
Sometimes when you hit Escape to bring up the menu, instead of blurring everything the entire screen tints blue/purple, looking in the logs one can find the following bad shader and error messages:

https://gist.github.com/gamax92/bb7ff4d3f15bb8b63b5d9a2307200c63

`GL_MAX_DUAL_SOURCE_DRAW_BUFFERS_EXT` only returns 1 and hence only `ps_out_0` and `ps_out_1` are created, yet this shader attempts to write to outputs 2 and 3 as well.

Removing the extra writes to outputs 2 and 3 results in the screen becoming textureless, dark, and brown.

Setting up extra outputs only results in nvidia complaining about `error: too many color outputs when using dual source output`

@awesie

**Comment by zekesonxx** [comment]
Do you have any way you know triggers it? Because I've literally *never* had this happen.

**Comment by gamax92** [comment]
I ... can't reproduce the issue any more :/ was going to get a screenshot of it.
It used to be that I could just go to Horizon, defender's spawn, stand in the middle of the room and hit escape. Eichenwalde's hero selection screen would also break and go completely purple

Oh well.

**Comment by gstark31897** [comment]
I might actually be able to reproduce this (only happens on the main menu though where the background tints blue/purple). (Ill attach a screenshot if I can get it to happen)

**Comment by gstark31897** [comment]
![screenshot_20170714_211738](https://user-images.githubusercontent.com/3507130/28235807-76835434-68db-11e7-9dee-36138632ef06.png)
Is this similar to the effect you're describing?

**Comment by gstark31897** [comment]
![screenshot_20170714_222710](https://user-images.githubusercontent.com/3507130/28236204-59930a5e-68e5-11e7-8bd7-bd325bfcc2df.png)

**Comment by zekesonxx** [comment]
@gstark31897 now that I've had happen tons of times.

I don't think this is what @gamax92 is referring to. If it's *not*, open a separate issue for that, because I'd love to track it down and fix it.

**Comment by gamax92** [comment]
Yeah, that's what I was referring

**Comment by gstark31897** [comment]
Yup, it happens pretty irregularly, but I get git it to happen within about 10-20 tries opening the game on the main menu. I've got a gtx 970 with the nvidia-dkms driver.

If you can't get it to happen locally, I'd be happy to  provide error logs or build/test code changes.

**Comment by zekesonxx** [comment]
So I have no idea how to read shader code (or even how shaders work) but I think we need a correctly compiled version to compare against.

**Comment by gstark31897** [comment]
It'd take some digging, but wine certainly has calls for compiling shaders at some point (I haven't personally looked through much of the code). We could probably add a debug statement to them to dump the binary shader that they produce, then see if it's actually something higher up that's causing it or if the binary is actually different when it bugs out.

**Comment by awesie** [comment]
My bet is that the code to determine whether to use dual-source blending or not for a particular shader might be wrong.

**Comment by gamax92** [comment]
@awesie you're probably right, did a dump of every shader made to a file and there's a shader identical to the one failing except that it didn't try to use dual-source blending for it.

**Comment by zekesonxx** [comment]
I think this might possibly be Overwatch's fault, at least partially.

The offending code is [right here](https://github.com/gamax92/wine-overwatch/blob/299018de8b10998f5ac47f1ae17a8961fe469fc9/dlls/wined3d/wined3d_private.h#L3814):

    #define IS_DUAL_SOURCE_BLEND(x) ((x) >= WINED3D_BLEND_SRC1COLOR && (x) <= WINED3D_BLEND_INVSRC1ALPHA)

What *is* Overwatch setting DestBlend to when it's correct and incorrect? [These](https://github.com/gamax92/wine-overwatch/blob/299018de8b10998f5ac47f1ae17a8961fe469fc9/include/wine/wined3d.h#L390) are the valid values (and their corresponding [D3D11](https://github.com/gamax92/wine-overwatch/blob/299018de8b10998f5ac47f1ae17a8961fe469fc9/include/d3d11.idl#L263) options), and only four (SRC1_COLOR, NV_SRC1_COLOR, SRC1_ALPHA, and INV_SRC1_ALPHA) trigger Wine to do dual source blending.

There's a slim chance the issue is really [here](https://github.com/gamax92/wine-overwatch/blob/299018de8b10998f5ac47f1ae17a8961fe469fc9/dlls/wined3d/state.c#L392), where Wine translates it's blend formats to OpenGL blend formats. But it sure looks correct to me.

Or [here](https://github.com/gamax92/wine-overwatch/blob/299018de8b10998f5ac47f1ae17a8961fe469fc9/dlls/d3d11/device.c#L1664), where Wine translates d3d11 blending info into wined3d blending info.

@gamax92 I'm not at my desktop (and I'm not burning all my mobile data to download OW), if you could redo your test and have it output the value of WINED3D_RS_DESTBLEND for both shaders, that would be helpful.

**Comment by gamax92** [comment]
Well, assuming those render states are of a `wined3d_blend` enum:

```
Good Shader:
    // WINED3D_RS_ALPHABLENDENABLE:  false
    // IS_DUAL_SOURCE_BLEND(WINED3D_RS_SRCBLENDALPHA = WINED3D_BLEND_ZERO):  false
    // IS_DUAL_SOURCE_BLEND(WINED3D_RS_DESTBLENDALPHA = WINED3D_BLEND_ZERO): false
    
    // IS_DUAL_SOURCE_BLEND(WINED3D_RS_SRCBLEND = WINED3D_BLEND_SRCALPHA):  false
    // IS_DUAL_SOURCE_BLEND(WINED3D_RS_DESTBLEND = WINED3D_BLEND_INVSRCALPHA): false

Bad Shader:
    // WINED3D_RS_ALPHABLENDENABLE:  false
    // IS_DUAL_SOURCE_BLEND(WINED3D_RS_SRCBLENDALPHA = WINED3D_BLEND_SRCALPHA):  false
    // IS_DUAL_SOURCE_BLEND(WINED3D_RS_DESTBLENDALPHA = WINED3D_BLEND_INVSRCALPHA): false
    
    // IS_DUAL_SOURCE_BLEND(WINED3D_RS_SRCBLEND = WINED3D_BLEND_ONE):  false
    // IS_DUAL_SOURCE_BLEND(WINED3D_RS_DESTBLEND = WINED3D_BLEND_SRC1COLOR): true
```

**Comment by zekesonxx** [comment]
... why is Overwatch using two dramatically different shader blend modes with the same shader?

@gamax92 not to doubt you but are you sure they're actually the same shader?

Alternatively, is Wine translating two different shaders into the "same" shader? Or trying to use new settings intended for a new shader with the previous one?

**Comment by awesie** [comment]
I was trying to look up what those blend states actually do if alpha blend enable is false. I think maybe IS_DUAL_SOURCE_BLEND should always return false if WINED3D_RS_ALPHABLENDENABLE is false?

**Comment by zekesonxx** [comment]
Also, msdn is really unclear:

> D3D11_BLEND_SRC1_COLOR
> The blend factor is data sources both as color data output by a pixel shader. There is no pre-blend operation. This blend factor **supports** dual-source color blending.

"supports"? Does that mean it *has* to be DSCB? Or is it just an option?

**Comment by zekesonxx** [comment]
@awesie well I mean you're the one who implemented it, so you're the most likely to know the answer.

**Comment by gamax92** [comment]
Makes sense, dx9's field was called `D3DRS_ALPHABLENDENABLE` hence `WINED3D_RS_ALPHABLENDENABLE`

dx10 and dx11 have `BlendEnable` in their `BLEND_DESC` though which ends up being assigned `WINED3D_RS_ALPHABLENDENABLE`, hence the odd misleading naming

https://msdn.microsoft.com/en-us/library/windows/desktop/ff476200(v=vs.85).aspx

Seems here that `BlendEnable` will enable/disable the usage of all of those `D3D11_BLEND` and `D3D11_BLEND_OP` fields, which should mean that yes `state_is_dual_source_blend` should always return false if there is no blending enabled.

I'll change the code to the following and test it out:
```
static inline BOOL state_is_dual_source_blend(const struct wined3d_state *state)
{
#define IS_DUAL_SOURCE_BLEND(x) ((x) >= WINED3D_BLEND_SRC1COLOR && (x) <= WINED3D_BLEND_INVSRC1ALPHA)
    return state->render_states[WINED3D_RS_ALPHABLENDENABLE] &&
            (IS_DUAL_SOURCE_BLEND(state->render_states[WINED3D_RS_SRCBLENDALPHA]) ||
             IS_DUAL_SOURCE_BLEND(state->render_states[WINED3D_RS_DESTBLENDALPHA]) ||
             IS_DUAL_SOURCE_BLEND(state->render_states[WINED3D_RS_SRCBLEND]) ||
             IS_DUAL_SOURCE_BLEND(state->render_states[WINED3D_RS_DESTBLEND]));
#undef IS_DUAL_SOURCE_BLEND
}
```

**Comment by zekesonxx** [comment]
if that works perhaps the field should be renamed to something like `WINED3D_RS_BLENDENABLE`.

**Comment by zekesonxx** [comment]
So I'm pretty sure this works. Using 2.12 with that version of state_is_dsb, I can't get the buggered out shader in 10+ game launches, and I got it in just 2 launches on normal 2.12.

I've played a few games and AFAICT there's no regressions caused by this anywhere.

**Comment by gamax92** [comment]
Alright, I'll add it into the repository, I took a look myself and couldn't find any regressions either.

**Comment by gamax92** [comment]
Closed by 6634c835a6774b264e294c6d135f745c8b907a9f
